File: experiments/problem1_nonperiodicity/analyze_period_two_residual_cylinder_minimizers.cpp
#include <algorithm>
#include <array>
#include <cstdint>
#include <iomanip>
#include <iostream>
#include <map>
#include <optional>
#include <sstream>
#include <stdexcept>
#include <string>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using u128 = unsigned __int128;

namespace {
constexpr int MAX_DEPTH = 16;
constexpr int MAX_RESIDUAL = 20;
constexpr std::array<unsigned,4> CHILD_MASK{0b1011,0b1100,0b1110,0b0011};

struct U128Hash { size_t operator()(u128 x) const noexcept { uint64_t lo=(uint64_t)x, hi=(uint64_t)(x>>64); uint64_t z=lo^(hi+0x9e3779b97f4a7c15ULL+(lo<<6)+(lo>>2)); z^=z>>30; z*=0xbf58476d1ce4e5b9ULL; z^=z>>27; z*=0x94d049bb133111ebULL; z^=z>>31; return (size_t)z; } };
std::string hex128(u128 x){ if(!x)return "0x0"; std::string s; while(x){ unsigned d=(unsigned)(x&15); s.push_back("0123456789abcdef"[d]); x>>=4;} std::reverse(s.begin(),s.end()); return "0x"+s; }
int bit_length(u128 x){int n=0;while(x){++n;x>>=1;}return n;}
u128 forward(char name,u128 s){u128 t=s^((s<<1)|(s<<2)); if(name=='t')return t; if(name=='u')return t^1; if(name=='p')return t^1^((s&1)==0?2:0); throw std::runtime_error("name");}
std::vector<u128> children(u128 s){std::array<u128,3>a{forward('t',s),forward('u',s),forward('p',s)}; std::sort(a.begin(),a.end()); return std::vector<u128>(a.begin(),std::unique(a.begin(),a.end()));}
u128 phase_start(char p){return p=='p'?3:1;} int expected_bits(char p,int k){return p=='p'?2*k:2*k-1;}
std::optional<u128> inverse_t(u128 out){if(out==0)return u128{0};int w=bit_length(out)-2;if(w<=0)return std::nullopt;u128 s=0;for(int i=0;i<w;++i){unsigned lo=0;if(i>=1)lo|=(unsigned)((s>>(i-1))&1);if(i>=2)lo|=(unsigned)((s>>(i-2))&1);unsigned b=(unsigned)((out>>i)&1)^lo;s|=(u128)b<<i;}if(forward('t',s)!=out)return std::nullopt;return s;}
std::optional<u128> inverse_g(char n,u128 out){if(n=='t')return inverse_t(out);std::optional<u128>s;if(n=='u')s=inverse_t(out^1);else {unsigned low=(unsigned)(out&1)^1; s=inverse_t(out^1^(low==0?2:0));} if(!s||forward(n,*s)!=out)return std::nullopt;return s;}
std::optional<u128> candidate(u128 q,int d){char g=d==0?'t':d==1?'u':'p';auto r=inverse_g(g,q);if(!r)return std::nullopt;return 4*(*r)+(unsigned)d;}

struct Key {char p; int k; u128 x; bool operator==(Key const&o)const{return p==o.p&&k==o.k&&x==o.x;}};
struct KeyHash {size_t operator()(Key const&a)const noexcept{return U128Hash{}(a.x)^((size_t)a.k<<1)^a.p;}};
std::unordered_map<Key,bool,KeyHash> memo;
std::map<std::pair<char,int>,std::vector<u128>> levels;
std::map<std::pair<char,int>,std::unordered_set<u128,U128Hash>> level_sets;
void build_levels(char p,int maxk){std::vector<u128> cur{phase_start(p)};levels[{p,1}]=cur;level_sets[{p,1}]={cur.begin(),cur.end()};for(int k=2;k<=maxk;++k){std::vector<u128> nxt; nxt.reserve(cur.size()*3);for(u128 x:cur){auto cs=children(x);nxt.insert(nxt.end(),cs.begin(),cs.end());}std::sort(nxt.begin(),nxt.end());nxt.erase(std::unique(nxt.begin(),nxt.end()),nxt.end());levels[{p,k}]=nxt;level_sets[{p,k}]={nxt.begin(),nxt.end()};cur.swap(nxt);}}
bool member(char p,int k,u128 x){if(k<1||bit_length(x)!=expected_bits(p,k))return false;if(k<=MAX_RESIDUAL){auto it=level_sets.find({p,k});return it!=level_sets.end()&&it->second.count(x);}Key key{p,k,x};auto it=memo.find(key);if(it!=memo.end())return it->second;u128 q=x>>2;int e=(int)(x&3);bool ok=false;for(int d=0;d<4&&!ok;++d){if(((CHILD_MASK[d]>>e)&1)==0)continue;auto par=candidate(q,d);if(par&&member(p,k-1,*par))ok=true;}memo.emplace(key,ok);return ok;}

uint64_t mask64(int w){return w==64?~0ULL:((uint64_t{1}<<w)-1);} // widths <=32 here
uint64_t inv_t_mod(uint64_t out,int w){uint64_t mask=w?mask64(w):0;out&=mask;uint64_t s=0;for(int i=0;i<w;++i){unsigned lo=0;if(i>=1)lo|=(s>>(i-1))&1;if(i>=2)lo|=(s>>(i-2))&1;unsigned b=((out>>i)&1)^lo;s|=(uint64_t)b<<i;}return s;}
uint64_t inv_g_mod(char n,uint64_t out,int w){uint64_t mask=w?mask64(w):0;out&=mask;if(n=='t')return inv_t_mod(out,w);if(n=='u')return inv_t_mod(out^1,w);unsigned low=(out&1)^1;return inv_t_mod(out^1^(low==0?2:0),w);}
uint64_t fringe(uint64_t s){uint64_t row=1+2*s;uint64_t odd=row^((row>>1)|(row>>2));return (odd<<1)^(odd|(odd>>1));}
std::string actual_driver(int L){uint64_t s=0;std::string out;for(int i=0;i<L;++i){out.push_back((s&3)==0?'u':'t');s=fringe(s);}return out;}
uint64_t survivor_word(std::string word){uint64_t s=0;int w=0;for(auto it=word.rbegin();it!=word.rend();++it){w+=2;int iw=w-2;s=inv_g_mod(*it,s,iw);s=inv_g_mod('p',s,iw);s=((s<<2)|3)&mask64(w);}return s;}

struct FilterResult {std::vector<u128> states; std::vector<size_t> funnel;};
FilterResult cylinder_filter(char p,int L,int k,uint64_t X){memo.clear();if(k<=L){std::vector<u128> s;if(member(p,k,X))s.push_back(X);return {s,{s.size()}};}int r=k-L;if(r<1||r>MAX_RESIDUAL)throw std::runtime_error("residual outside campaign");std::vector<u128> cur=levels.at({p,r});std::vector<size_t> funnel{cur.size()};for(int idx=L-1;idx>=0;--idx){int e=(X>>(2*idx))&3;int c=r+(L-1-idx);std::vector<u128> nxt;nxt.reserve(cur.size());for(u128 q:cur){u128 y=(q<<2)|(unsigned)e;if(member(p,c+1,y))nxt.push_back(y);}cur.swap(nxt);funnel.push_back(cur.size());if(cur.empty())break;}return {cur,funnel};}

int kappa(char p,int L,uint64_t X,int lower,int upper,FilterResult* final=nullptr){for(int k=lower;k<=upper;++k){auto row=cylinder_filter(p,L,k,X);if(!row.states.empty()){if(final)*final=std::move(row);return k;}}return -1;}
std::string funnel_string(std::vector<size_t> const&v){std::ostringstream o;for(size_t i=0;i<v.size();++i){if(i)o<<',';o<<v[i];}return o.str();}

uint64_t small_exhaustive(){uint64_t checks=0;for(char p:{'p','u'}){for(int k=1;k<=10;++k){auto const& full=levels.at({p,k});for(int L=1;L<=std::min(5,k+1);++L){uint64_t mod=uint64_t{1}<<(2*L);std::map<uint64_t,size_t> direct;for(u128 x:full)direct[(uint64_t)x&(mod-1)]++;for(uint64_t X=0;X<mod;++X){auto f=cylinder_filter(p,L,k,X);size_t want=direct[X];if(f.states.size()!=want){std::cerr<<"mismatch "<<p<<" k"<<k<<" L"<<L<<" X"<<X<<" got"<<f.states.size()<<" want"<<want<<"\n";throw std::runtime_error("small mismatch");}++checks;}}}}return checks;}
}

int main(){build_levels('p',MAX_RESIDUAL);build_levels('u',MAX_RESIDUAL);const uint64_t small_checks=small_exhaustive();
std::map<int,std::pair<int,int>> expected{{1,{1,2}},{2,{3,2}},{3,{7,2}},{4,{8,7}},{5,{8,12}},{6,{12,14}},{7,{13,14}},{8,{17,14}},{9,{17,18}},{10,{17,19}},{11,{21,26}},{12,{28,27}},{13,{30,30}},{14,{33,30}},{15,{34,30}},{16,{36,30}}};
int prevp=1,prevu=1;uint64_t checksum=1469598103934665603ULL;for(int L=1;L<=MAX_DEPTH;++L){uint64_t X=survivor_word(actual_driver(L));FilterResult fp,fu;int kp=kappa('p',L,X,1,expected[L].first,&fp);int ku=kappa('u',L,X,1,expected[L].second,&fu);if(kp!=expected[L].first||ku!=expected[L].second)throw std::runtime_error("kappa mismatch");std::cout<<"actual depth="<<L<<" residue=0x"<<std::hex<<X<<std::dec<<" kappa_p="<<kp<<" p_minimizers="<<fp.states.size()<<" p_funnel="<<funnel_string(fp.funnel)<<" kappa_u="<<ku<<" u_minimizers="<<fu.states.size()<<" u_funnel="<<funnel_string(fu.funnel)<<"\n";checksum^=X+((uint64_t)kp<<40)+((uint64_t)ku<<48);checksum*=1099511628211ULL;prevp=kp;prevu=ku;}
for(auto [word,label]:std::vector<std::pair<std::string,std::string>>{{"tutututttutu","w12"},{"tutututttututu","w14"},{"tutututttutututu","w16"}}){int L=word.size();uint64_t X=survivor_word(word);auto f=cylinder_filter('u',L,25,X);if(f.states.size()!=1)throw std::runtime_error("counter unique");std::cout<<"counter "<<label<<" depth="<<L<<" residue=0x"<<std::hex<<X<<std::dec<<" minimizers="<<f.states.size()<<" state="<<hex128(f.states.front())<<" funnel="<<funnel_string(f.funnel)<<"\n";}
std::cout<<"small_exhaustive_cylinder_checks="<<small_checks<<"\n";std::cout<<"max_depth="<<MAX_DEPTH<<"\n";std::cout<<"max_residual="<<MAX_RESIDUAL<<"\n";std::cout<<"p_residual_states="<<levels[{ 'p',MAX_RESIDUAL}].size()<<"\n";std::cout<<"u_residual_states="<<levels[{ 'u',MAX_RESIDUAL}].size()<<"\n";std::cout<<"checksum=0x"<<std::hex<<checksum<<std::dec<<"\n";
}

File: experiments/problem1_nonperiodicity/analyze_period_two_residual_cylinder_minimizers.py
#!/usr/bin/env python3
"""Exact residual-frontier factorization for phase survivor cylinders.

For phase a, complexity k, depth L, and residue X modulo 4**L, define

    C_(a,k)(X,L) = {x in O_(a,k): x == X (mod 4**L)}.

When k>L, projection by 2L bits injects this set into O_(a,k-L), and

    C_(a,k)(X,L)
      = {X + 4**L h : h in O_(a,k-L), X + 4**L h in O_(a,k)}.

The final membership condition is decided by the exact recursive lift criterion.
Reading the fixed base-four digits of X from high to low filters the residual
frontier without constructing the full complexity-k frontier.
"""
from __future__ import annotations

import argparse
import hashlib
import json
from functools import lru_cache
from typing import Any

PHASES = ("p", "u")
CHILD_DIGIT_MASK = {0: 0b1011, 1: 0b1100, 2: 0b1110, 3: 0b0011}
DEFAULT_MAXIMUM_DEPTH = 10
ABSOLUTE_MAXIMUM_DEPTH = 12
DEFAULT_MAXIMUM_RESIDUAL = 13
ABSOLUTE_MAXIMUM_RESIDUAL = 16
EXPECTED = {
    1: {"p": 1, "u": 2}, 2: {"p": 3, "u": 2},
    3: {"p": 7, "u": 2}, 4: {"p": 8, "u": 7},
    5: {"p": 8, "u": 12}, 6: {"p": 12, "u": 14},
    7: {"p": 13, "u": 14}, 8: {"p": 17, "u": 14},
    9: {"p": 17, "u": 18}, 10: {"p": 17, "u": 19},
    11: {"p": 21, "u": 26}, 12: {"p": 28, "u": 27},
}
COUNTER_WORDS = {
    "w12": "tutututttutu",
    "w14": "tutututttututu",
    "w16": "tutututttutututu",
}
KNOWN_COUNTEREXAMPLE = 0x1BCD3A7B3FDFB


class ResidualCylinderLimitError(RuntimeError):
    """Raised before a controlled residual-frontier campaign exceeds its cap."""


def forward_generator(name: str, state: int) -> int:
    if state < 0:
        raise ValueError("state must be nonnegative")
    stepped = state ^ ((state << 1) | (state << 2))
    if name == "t":
        return stepped
    if name == "u":
        return stepped ^ 1
    if name == "p":
        return stepped ^ 1 ^ (2 if state & 1 == 0 else 0)
    raise ValueError("unknown generator")


def frontier_children(state: int) -> tuple[int, ...]:
    return tuple(sorted({forward_generator(name, state) for name in "tup"}))


def phase_start(phase: str) -> int:
    if phase == "p":
        return 3
    if phase == "u":
        return 1
    raise ValueError("phase must be p or u")


def expected_bit_length(phase: str, complexity: int) -> int:
    if complexity <= 0:
        raise ValueError("complexity must be positive")
    return 2 * complexity if phase == "p" else 2 * complexity - 1


def inverse_t(output: int) -> int | None:
    if output < 0:
        raise ValueError("output must be nonnegative")
    if output == 0:
        return 0
    width = output.bit_length() - 2
    if width <= 0:
        return None
    state = 0
    for position in range(width):
        lower = 0
        if position >= 1:
            lower |= (state >> (position - 1)) & 1
        if position >= 2:
            lower |= (state >> (position - 2)) & 1
        state |= ((((output >> position) & 1) ^ lower) << position)
    return state if forward_generator("t", state) == output else None


def inverse_generator(name: str, output: int) -> int | None:
    if name == "t":
        state = inverse_t(output)
    elif name == "u":
        state = inverse_t(output ^ 1)
    elif name == "p":
        recovered_low_bit = (output & 1) ^ 1
        state = inverse_t(output ^ 1 ^ (2 if recovered_low_bit == 0 else 0))
    else:
        raise ValueError("unknown generator")
    if state is None or forward_generator(name, state) != output:
        return None
    return state


def candidate_parent(quotient: int, parent_digit: int) -> int | None:
    if quotient < 0 or parent_digit not in range(4):
        raise ValueError("invalid candidate-parent argument")
    name = "t" if parent_digit == 0 else "u" if parent_digit == 1 else "p"
    residual = inverse_generator(name, quotient)
    return None if residual is None else 4 * residual + parent_digit


def build_phase_levels(phase: str, maximum_complexity: int) -> list[set[int]]:
    levels: list[set[int]] = [set(), {phase_start(phase)}]
    for _ in range(2, maximum_complexity + 1):
        levels.append({child for state in levels[-1] for child in frontier_children(state)})
    return levels


def make_member(levels: dict[str, list[set[int]]], base_cap: int):
    @lru_cache(maxsize=None)
    def member(phase: str, complexity: int, target: int) -> bool:
        if complexity < 1 or target < 0:
            return False
        if target.bit_length() != expected_bit_length(phase, complexity):
            return False
        if complexity <= base_cap:
            return target in levels[phase][complexity]
        quotient, digit = divmod(target, 4)
        for parent_digit in range(4):
            if not (CHILD_DIGIT_MASK[parent_digit] >> digit) & 1:
                continue
            parent = candidate_parent(quotient, parent_digit)
            if parent is not None and member(phase, complexity - 1, parent):
                return True
        return False
    return member


def inverse_t_mod(output: int, width: int) -> int:
    mask = (1 << width) - 1 if width else 0
    output &= mask
    state = 0
    for position in range(width):
        lower = 0
        if position >= 1:
            lower |= (state >> (position - 1)) & 1
        if position >= 2:
            lower |= (state >> (position - 2)) & 1
        state |= ((((output >> position) & 1) ^ lower) << position)
    return state


def inverse_generator_mod(name: str, output: int, width: int) -> int:
    mask = (1 << width) - 1 if width else 0
    output &= mask
    if name == "t":
        return inverse_t_mod(output, width)
    if name == "u":
        return inverse_t_mod(output ^ 1, width)
    if name == "p":
        recovered_low_bit = (output & 1) ^ 1
        return inverse_t_mod(
            output ^ 1 ^ (2 if recovered_low_bit == 0 else 0), width
        )
    raise ValueError("unknown generator")


def fringe_step(state: int) -> int:
    row = 1 + 2 * state
    odd = row ^ ((row >> 1) | (row >> 2))
    return (odd << 1) ^ (odd | (odd >> 1))


def actual_driver(length: int) -> str:
    state = 0
    word: list[str] = []
    for _ in range(length):
        word.append("u" if state & 3 == 0 else "t")
        state = fringe_step(state)
    return "".join(word)


def survivor_for_word(word: str) -> int:
    state = 0
    precision = 0
    for branch in reversed(word):
        precision += 2
        inner_width = precision - 2
        state = inverse_generator_mod(branch, state, inner_width)
        state = inverse_generator_mod("p", state, inner_width)
        state = ((state << 2) | 3) & ((1 << precision) - 1)
    return state


def cylinder_filter(
    *, phase: str, depth: int, complexity: int, residue: int,
    levels: dict[str, list[set[int]]], member,
) -> tuple[list[int], list[int]]:
    if depth <= 0 or complexity <= 0:
        raise ValueError("depth and complexity must be positive")
    if not 0 <= residue < 1 << (2 * depth):
        raise ValueError("residue outside the cylinder modulus")
    member.cache_clear()
    if complexity <= depth:
        states = [residue] if member(phase, complexity, residue) else []
        return states, [len(states)]
    residual_complexity = complexity - depth
    if residual_complexity >= len(levels[phase]):
        raise ResidualCylinderLimitError("residual complexity exceeds built frontier")
    current = sorted(levels[phase][residual_complexity])
    funnel = [len(current)]
    for index in range(depth - 1, -1, -1):
        digit = (residue >> (2 * index)) & 3
        next_complexity = residual_complexity + (depth - index)
        current = [
            4 * quotient + digit
            for quotient in current
            if member(phase, next_complexity, 4 * quotient + digit)
        ]
        funnel.append(len(current))
        if not current:
            break
    return current, funnel


def cylinder_kappa(
    *, phase: str, depth: int, residue: int, maximum_complexity: int,
    levels: dict[str, list[set[int]]], member,
) -> tuple[int, list[int], list[int]]:
    for complexity in range(1, maximum_complexity + 1):
        states, funnel = cylinder_filter(
            phase=phase, depth=depth, complexity=complexity, residue=residue,
            levels=levels, member=member,
        )
        if states:
            return complexity, states, funnel
    raise AssertionError("maximum complexity did not reach the cylinder")


def exhaustive_small_check(levels: dict[str, list[set[int]]], member) -> int:
    checks = 0
    for phase in PHASES:
        for complexity in range(1, 8):
            full = levels[phase][complexity]
            for depth in range(1, min(4, complexity + 1) + 1):
                modulus = 1 << (2 * depth)
                direct: dict[int, int] = {}
                for state in full:
                    direct[state % modulus] = direct.get(state % modulus, 0) + 1
                for residue in range(modulus):
                    states, _ = cylinder_filter(
                        phase=phase, depth=depth, complexity=complexity,
                        residue=residue, levels=levels, member=member,
                    )
                    if len(states) != direct.get(residue, 0):
                        raise AssertionError("residual factorization disagreed with direct frontier")
                    checks += 1
    return checks


def run_campaign(
    maximum_depth: int = DEFAULT_MAXIMUM_DEPTH,
    maximum_residual: int = DEFAULT_MAXIMUM_RESIDUAL,
) -> dict[str, Any]:
    if not 1 <= maximum_depth <= ABSOLUTE_MAXIMUM_DEPTH:
        raise ResidualCylinderLimitError("maximum depth outside controlled range")
    if not 1 <= maximum_residual <= ABSOLUTE_MAXIMUM_RESIDUAL:
        raise ResidualCylinderLimitError("maximum residual outside controlled range")
    levels = {
        phase: build_phase_levels(phase, maximum_residual) for phase in PHASES
    }
    member = make_member(levels, maximum_residual)
    small_checks = exhaustive_small_check(levels, member)

    rows: list[dict[str, Any]] = []
    for depth in range(1, maximum_depth + 1):
        residue = survivor_for_word(actual_driver(depth))
        phase_rows: dict[str, Any] = {}
        for phase in PHASES:
            expected = EXPECTED[depth][phase]
            complexity, states, funnel = cylinder_kappa(
                phase=phase, depth=depth, residue=residue,
                maximum_complexity=expected, levels=levels, member=member,
            )
            if complexity != expected:
                raise AssertionError("actual cylinder kappa mismatch")
            phase_rows[phase] = {
                "minimum_complexity": complexity,
                "residual_complexity": max(0, complexity - depth),
                "minimum_state_count": len(states),
                "funnel": funnel,
                "example_state_hex": hex(states[0]),
            }
        rows.append({
            "depth": depth,
            "actual_word": actual_driver(depth),
            "residue": residue,
            "residue_hex": hex(residue),
            "phases": phase_rows,
        })

    counterexamples: dict[str, Any] = {}
    for label, word in COUNTER_WORDS.items():
        depth = len(word)
        residue = survivor_for_word(word)
        states, funnel = cylinder_filter(
            phase="u", depth=depth, complexity=25, residue=residue,
            levels=levels, member=member,
        )
        if states != [KNOWN_COUNTEREXAMPLE]:
            raise AssertionError("counterexample cylinder was not uniquely minimized")
        counterexamples[label] = {
            "depth": depth,
            "word": word,
            "residue_hex": hex(residue),
            "complexity": 25,
            "unique_minimizer_hex": hex(states[0]),
            "funnel": funnel,
        }

    payload: dict[str, Any] = {
        "status": "exact-residual-factorization-and-controlled-minimization",
        "maximum_depth": maximum_depth,
        "maximum_residual_complexity": maximum_residual,
        "small_exhaustive_cylinder_checks": small_checks,
        "theorem": (
            "For k>L, the map x -> x>>(2L) injects the phase-cylinder "
            "intersection into O_(a,k-L), and the intersection is exactly the "
            "high-to-low recursive lift filter over the fixed base-four digits."
        ),
        "actual_rows": rows,
        "counterexample_cylinders": counterexamples,
        "scientific_boundary": (
            "The factorization decides finite cylinders and compresses their exact "
            "search. It does not prove unbounded actual phase complexity, recurring "
            "positive return penalties, exclusion of eventual period two, or Rule 30 "
            "center nonperiodicity."
        ),
    }
    encoded = json.dumps(payload, sort_keys=True, separators=(",", ":")).encode()
    payload["certificate_sha256"] = hashlib.sha256(encoded).hexdigest()
    return payload


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--maximum-depth", type=int, default=DEFAULT_MAXIMUM_DEPTH)
    parser.add_argument(
        "--maximum-residual", type=int, default=DEFAULT_MAXIMUM_RESIDUAL
    )
    args = parser.parse_args()
    print(json.dumps(
        run_campaign(args.maximum_depth, args.maximum_residual),
        indent=2, sort_keys=True,
    ))


if __name__ == "__main__":
    main()
